### PR TITLE
[Segment Replication] Fix bug of replica shard's translog not purging on index flush.

### LINKED ISSUE #4927: [BUG] [Segment Replication] Replica shard's translog is not getting purged after flush on index
**Describe the bug**
With Segment Replication enabled when a flush is done on index, replica shards are not purging their translog and replica's translog size is increasing with new segment files. This can lead to disk running out of space on node very quickly. This only happens with replica shards, primary shards do purge their translog on a flush.

**To Reproduce**
-> Run a cluster and create an index with segment replication enabled.
-> Insert few docs, then flush on the index.
-> Compare the translog size on disk  of primary and replica.

**Expected behavior**
Same as primary shards replica shards should also purge their translog after a flush on index.

### PR DESCRIPTION
**Pull request by Rishikesh1159**
Signed-off-by: Rishikesh1159 <rishireddy1159@gmail.com>

### Description
This PR fixes bug of replica shard's translog not purging on index flush when segment replication is enabled and fixes related unit tests.

-> Before this fix, we were not updating local checkpoint of safe commit on the replica's translog deletion policy. Updating this on a different incoming generation purges the translog on index flush.

->  This PR updates the `testReplicaReceivesGenIncrease()` because previously this test was only asserting doc counts on shards, which is not enough to make sure that replicas translog has been purged. So, we update this test to assert on few translog stats to make sure replicas translog has been purged.

-> This PR also updates the `testNRTReplicaPromotedAsPrimary()` because assert on totalDocs would fail as at that point we will only have newly added additionalDocs in our translog. Previous operations would be commited and purged from translog before the assert.

### Issues Resolved
Resolves #4927 

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -168,6 +168,7 @@ Inspired from [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 - [BUG]: flaky test index/80_geo_point/Single point test([#4860](https://github.com/opensearch-project/OpenSearch/pull/4860))
 - Fix bug in SlicedInputStream with zero length ([#4863](https://github.com/opensearch-project/OpenSearch/pull/4863))
 - Fix a bug on handling an invalid array value for point type field #4900([#4900](https://github.com/opensearch-project/OpenSearch/pull/4900))
+- Fix bug of replica shard's translog not purging on index flush.([4928](https://github.com/opensearch-project/OpenSearch/pull/4928))
```
Comment (mch2) [inline CHANGELOG.md:171]: please add context that this only happened with segment replication enabled.

**Review by mch2 — APPROVED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (dblock) [reply CHANGELOG.md:171]: remove the period in the sentence too

fix the line above as well, remove the extra #4900 while you're at it?

**Review by gbbafna — APPROVED** [review]
LGTM .

**Review by Bukhtawar — APPROVED** [review]
Do we need integ tests for this?

**Comment by Rishikesh1159** [comment]
> Do we need integ tests for this?

We already have `testReplicaReceivesGenIncrease()` which is very close/similar to an integ test. Even if we write an integ test we would be asserting on same stats and would be repetition of same existing test. This test `testReplicaReceivesGenIncrease()` is working fine and catching the bug correctly, so I think we might not need an integ test. But if you have other thoughts/concerns I can definitely add an integ test.

**Pull request merged**